File: python/leetcode/208.py
class TrieNode:
    def __init__(self):
        self.word = False
        self.child = {}

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.child:
                node.child[ch] = TrieNode()
            node = node.child[ch]
        node.word = True

    def search(self, word):
        node = self.root
        for ch in word:
            if ch not in node.child: return False
            node = node.child[ch]
        return node.word

    def startsWith(self, prefix):
        node = self.root
        for ch in prefix:
            if ch not in node.child: return False
            node = node.child[ch]
        return True

File: cpp/leetcode/208.cpp
class TrieNode {
public:
    TrieNode *child[26];
    bool isWord;
    TrieNode(){
        isWord = false;
        for (auto &ptr : child) ptr = NULL; // necessary
    }
};

class Trie {
public:
    Trie() { root = new TrieNode();}

    void insert(string word) {
        TrieNode *ptr = root;
        for (auto ch : word) {
            int i = ch - 'a';
            if (!ptr->child[i]) ptr->child[i] = new TrieNode();
            ptr = ptr->child[i];
        }
        ptr->isWord = true;
    }

    bool search(string word) {
        TrieNode *ptr = root;
        for (auto ch : word) {
            int i = ch - 'a';
            if (!ptr->child[i]) return false;
            ptr = ptr->child[i];
        }
        return ptr->isWord;
    }

    bool startsWith(string prefix) {
        TrieNode *ptr = root;
        for (auto ch : prefix) {
            int i = ch - 'a';
            if (!ptr->child[i]) return false;
            ptr = ptr->child[i];
        }
        return true;
    }

private:
    TrieNode* root;
};